js/ui/codingManager.js

```diff
@@ -29,239 +46,525 @@ function _isBuilderSpeedwagon(window) {
+    let {
+        button,
+        rect,
+        startAngle,
+        finishAngle,
+        startOpacity,
+        finishOpacity,
+        opacityDelay,
+        onRotationComplete,
+        onButtonFadeComplete
+    } = props;
+
+    // this API is deprecated but the best option here to
+    // animate around a point outside of the actor
+    button.rotation_center_y = new Clutter.Vertex({
+        x: button.width - (rect.width * 0.5),
+        y: button.height * 0.5,
+        z: 0
+    });
+    button.rotation_angle_y = startAngle;
+    button.opacity = startOpacity;
+    Tweener.addTween(button, {
+        rotation_angle_y: finishAngle,
+        time: WINDOW_ANIMATION_TIME * 4,
+        transition: 'easeOutQuad',
+        onComplete: function() {
+            if (onRotationComplete) {
+                onRotationComplete();
+            }
+        }
+    });
+    Tweener.addTween(button, {
+        opacity: finishOpacity,
+        time: WINDOW_ANIMATION_TIME * 2,
+        transition: 'linear',
+        delay: opacityDelay,
+        onComplete: function() {
+            if (onButtonFadeComplete) {
+                onButtonFadeComplete();
+            }
+        }
+    });
+}
 
-        if (!this._isBuilder(window.get_flatpak_id()) &&
-            !isSpeedwagonForBuilder)
-            return false;
+const WindowTrackingButton = new Lang.Class({
+    Name: 'WindowTrackingButton',
+    Extends: GObject.Object,
+    Properties: {
+        window: GObject.ParamSpec.object('window',
+                                         '',
+                                         '',
+                                         GObject.ParamFlags.READWRITE |
+                                         GObject.ParamFlags.CONSTRUCT_ONLY,
+                                         Meta.Window),
+        builder_window: GObject.ParamSpec.object('builder-window',
+                                                 '',
+                                                 '',
+                                                 GObject.ParamFlags.READWRITE,
+                                                 Meta.Window)
+    },
+    Signals: {
+        'clicked': {}
+    },
+
+    _init: function(params) {
+        this.parent(params);
+
+        let rect = this.window.get_frame_rect();
+
+        // The button will be auto-added to the manager. Note that in order to
+        // remove this button, you will need to call eject() from outside
+        // this class or use removeChrome from within it.
+        this._button = _createViewSourceButtonInRectCorner(rect);
+
+        // Do required setup to make this button interactive, by allowing
+        // it to be focused, connecting to the button-press-event signal
+        // and adding chrome to the toplevel layoutManager.
+        this._button.reactive = true;
+        this._button.can_focus = true;
+        this._button.track_hover = true;
+        this._button.connect('button-press-event', Lang.bind(this, function() {
```

Comment:
I believe the button has a `clicked` signal too; this event will also be emitted for e.g. right clicks.